File: src/lib/mock/progress.ts
import type {
  ActivityDay,
  DashboardData,
  PatternStat,
  Progress,
  ProblemSummary,
} from "@/src/lib/types";
import { PATTERNS } from "@/src/lib/types";
import { MOCK_PROBLEMS, MOCK_STATUSES } from "./problems";

export const MOCK_PROGRESS: Progress = {
  solved: 42,
  total: 150,
  attempted: 61,
  streakDays: 7,
  bestStreakDays: 12,
  mastered: 18,
  needsReview: 12,
};

/** Pattern totals shown on the dashboard and the library rail. */
export const MOCK_PATTERN_STATS: PatternStat[] = [
  { pattern: "Arrays & Hashing", solved: 8, total: 8 },
  { pattern: "Two Pointers", solved: 6, total: 7 },
  { pattern: "Sliding Window", solved: 5, total: 6 },
  { pattern: "Stack", solved: 4, total: 7 },
  { pattern: "Binary Search", solved: 3, total: 7 },
  { pattern: "Linked List", solved: 2, total: 6 },
  { pattern: "Trees", solved: 5, total: 12 },
  { pattern: "Heap / Priority Queue", solved: 1, total: 5 },
  { pattern: "Backtracking", solved: 0, total: 6 },
  { pattern: "Graphs", solved: 1, total: 8 },
  { pattern: "1-D DP", solved: 2, total: 10 },
  { pattern: "2-D DP", solved: 0, total: 7 },
  { pattern: "Greedy", solved: 1, total: 5 },
  { pattern: "Intervals", solved: 2, total: 4 },
  { pattern: "Bit Manipulation", solved: 3, total: 5 },
];

/** Deterministic ~26 weeks of daily activity ending today. */
function buildActivity(): ActivityDay[] {
  const days: ActivityDay[] = [];
  const today = new Date();
  const totalDays = 26 * 7;
  let seed = 9;
  const rnd = () => {
    seed = (seed * 1103515245 + 12345) & 0x7fffffff;
    return seed / 0x7fffffff;
  };
  for (let i = totalDays - 1; i >= 0; i--) {
    const d = new Date(today);
    d.setDate(today.getDate() - i);
    const r = rnd();
    let count = r < 0.42 ? 0 : r < 0.62 ? 1 : r < 0.8 ? 2 : r < 0.92 ? 3 : 5;
    // taper the most recent weeks like the mockup
    if (i < 14) count = Math.min(count, r < 0.6 ? 1 : 2);
    days.push({ date: d.toISOString().slice(0, 10), count });
  }
  return days;
}

export function buildDashboard(): DashboardData {
  const summaries = problemSummaries();
  const inProgress = summaries.find((p) => p.status === "in-progress");
  const sorted = [...MOCK_PATTERN_STATS].sort(
    (a, b) => a.solved / a.total - b.solved / b.total
  );
  const months = new Date();
  const start = new Date(months);
  start.setDate(start.getDate() - 26 * 7);
  const mid = new Date(months);
  mid.setDate(mid.getDate() - 13 * 7);
  const monthName = (d: Date) => d.toLocaleString("en", { month: "short" });

  return {
    progress: MOCK_PROGRESS,
    activity: buildActivity(),
    cumulative: [0, 1, 3, 6, 9, 12, 16, 20, 25, 30, 35, 42],
    axisLabels: [monthName(start), monthName(mid)],
    focus: sorted.slice(0, 3),
    strong: sorted.slice(-3).reverse(),
    continueProblem: inProgress,
    patternStats: PATTERNS.map(
      (pattern) =>
        MOCK_PATTERN_STATS.find((s) => s.pattern === pattern) ?? {
          pattern,
          solved: 0,
          total: 0,
        }
    ),
  };
}

export function problemSummaries(): ProblemSummary[] {
  return MOCK_PROBLEMS.map((p) => {
    const s = MOCK_STATUSES[p.id] ?? { status: "todo" as const };
    return {
      id: p.id,
      number: p.number,
      title: p.title,
      pattern: p.pattern,
      difficulty: p.difficulty,
      source: p.source,
      status: s.status,
      lastAttempted: s.lastAttempted,
    };
  });
}

File: src-tauri/src/domain/progress.rs
//! Progress + dashboard types — camelCase like `src/lib/types.ts`
//! (`streakDays`, `needsReview`, …). Dashboard types are filled by task
//! 0009; `Progress` ships with 0008.

use serde::{Deserialize, Serialize};

use crate::domain::problem::{Pattern, ProblemStatus, ProblemSummary};

/// UI status mutations (`set_problem_status`). Commands don't trust
/// callers: `MarkMastered` is rejected unless the problem is solved.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StatusAction {
    MarkMastered,
    UnmarkMastered,
    NeedsReview,
    ClearReview,
}

/// Where a lesson sits for this user (LESSON_COURSE_DESIGN.md §6.4). Mirrors
/// the `lesson_progress.status` CHECK constraint and the TS `LessonStatus`.
/// `NotStarted` is the absence of a stored row, surfaced by the UI only.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum LessonStatus {
    NotStarted,
    InProgress,
    Complete,
}

impl LessonStatus {
    /// Reads the stored `status` text back into the enum; an unrecognized
    /// value (only possible via out-of-band DB edits) reads as `NotStarted`.
    pub fn from_wire(s: &str) -> Self {
        match s {
            "in-progress" => Self::InProgress,
            "complete" => Self::Complete,
            _ => Self::NotStarted,
        }
    }
}

/// One lesson's stored progress row (`record_lesson_progress` /
/// `get_lesson_progress`). Timestamps are local-time ISO-8601 strings, like
/// every other stamp in the DB.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LessonProgress {
    pub lesson_id: String,
    pub unit_id: String,
    pub status: LessonStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub started_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<String>,
}

/// Per-problem user state for the workspace: bookmark icon, mastered flag,
/// and the code snapshot from the most recent run/submit so the editor can
/// restore where the user left off.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProblemUserState {
    pub status: ProblemStatus,
    pub bookmarked: bool,
    pub mastered: bool,
    #[serde(rename = "lastCode", skip_serializing_if = "Option::is_none")]
    pub last_code: Option<String>,
    #[serde(rename = "lastLanguage", skip_serializing_if = "Option::is_none")]
    pub last_language: Option<crate::domain::run::Language>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Progress {
    pub solved: u32,
    pub total: u32,
    pub attempted: u32,
    pub streak_days: u32,
    pub best_streak_days: u32,
    pub mastered: u32,
    pub needs_review: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ActivityDay {
    /// ISO date (yyyy-mm-dd).
    pub date: String,
    pub count: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PatternStat {
    pub pattern: Pattern,
    pub solved: u32,
    pub total: u32,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DashboardData {
    pub progress: Progress,
    /// Daily solve counts for the heatmap, oldest first (~26 weeks).
    pub activity: Vec<ActivityDay>,
    /// Cumulative solved counts for the progress line, oldest first.
    pub cumulative: Vec<u32>,
    /// Month labels for the line chart axis: [start, mid].
    pub axis_labels: [String; 2],
    pub focus: Vec<PatternStat>,
    pub strong: Vec<PatternStat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub continue_problem: Option<ProblemSummary>,
    pub pattern_stats: Vec<PatternStat>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn progress_round_trips_with_camel_case_fields() {
        let value = json!({
            "solved": 3, "total": 12, "attempted": 5,
            "streakDays": 2, "bestStreakDays": 4,
            "mastered": 1, "needsReview": 1
        });
        let parsed: Progress = serde_json::from_value(value.clone()).unwrap();
        assert_eq!(serde_json::to_value(parsed).unwrap(), value);
    }

    #[test]
    fn problem_user_state_round_trips_and_omits_absent_snapshot() {
        let value = json!({
            "status": "solved",
            "bookmarked": true,
            "mastered": false,
            "lastCode": "def solve(a, b):\n    return a + b",
            "lastLanguage": "python"
        });
        let parsed: ProblemUserState = serde_json::from_value(value.clone()).unwrap();
        assert_eq!(serde_json::to_value(&parsed).unwrap(), value);

        let fresh = ProblemUserState {
            status: ProblemStatus::Todo,
            bookmarked: false,
            mastered: false,
            last_code: None,
            last_language: None,
        };
        let v = serde_json::to_value(&fresh).unwrap();
        assert!(v.get("lastCode").is_none());
        assert!(v.get("lastLanguage").is_none());
    }

    #[test]
    fn dashboard_data_round_trips() {
        let value = json!({
            "progress": {
                "solved": 1, "total": 12, "attempted": 1,
                "streakDays": 1, "bestStreakDays": 1,
                "mastered": 0, "needsReview": 0
            },
            "activity": [{ "date": "2026-06-12", "count": 1 }],
            "cumulative": [0, 1],
            "axisLabels": ["Dec", "Mar"],
            "focus": [{ "pattern": "Stack", "solved": 0, "total": 2 }],
            "strong": [{ "pattern": "Graphs", "solved": 1, "total": 2 }],
            "continueProblem": {
                "id": "x", "number": 1, "title": "X", "pattern": "Stack",
                "difficulty": "Easy", "source": "built-in", "status": "in-progress"
            },
            "patternStats": [{ "pattern": "Stack", "solved": 0, "total": 2 }]
        });
        let parsed: DashboardData = serde_json::from_value(value.clone()).unwrap();
        assert_eq!(serde_json::to_value(parsed).unwrap(), value);
    }
}
